src/range.rs

```diff
@@ -342,10 +342,79 @@ fn group_adjacent_locations(
 impl<V: Ord + Clone> Range<V> {
     /// Computes the union of this `Range` and another.
     pub fn union(&self, other: &Self) -> Self {
-        self.complement()
-            .intersection(&other.complement())
-            .complement()
-            .check_invariants()
+        let end_before_start_exclusive = |end: &Bound<_>, start: &Bound<_>| match (end, start) {
```

Comment:
This looks similar to the `valid_segment` helper. We should probably put the helpers together and document how they differ.